sys/kern/event.c

```diff
@@ -1,15 +1,452 @@
+ * events via `eventlist`. If there are no events available, blocks for at
+ * most `tsp`. If `tsp` is null, blocks infinitely. If the timeout is not
+ * positive, returns immediately.
+ */
+static int kqueue_scan(kqueue_t *kq, kevent_t *eventlist, size_t nevents,
+                       timespec_t *tsp, int *retval) {
+  int error, event, timeout;
+  size_t count = 0;
+  systime_t sleepts;
+  knote_tailq_t knqueue;
+  knote_t *kn;
+
+  TAILQ_INIT(&knqueue);
+
+  if (tsp) {
+    timeout = ts2hz(tsp);
+    sleepts = getsystime() + timeout;
+    if (timeout <= 0)
+      timeout = -1; /* don't block */
+  } else {
+    timeout = 0; /* no timeout, wait forever */
+  }
+
+  mtx_lock(&kq->kq_lock);
+
+  /* Block until there are no events or we time out. */
+  while (kq->kq_count == 0) {
+    if (timeout < 0) {
+      error = 0;
+      goto done;
+    }
+
+    error = cv_wait_timed(&kq->kq_cv, &kq->kq_lock, timeout);
+    if (error == EINTR) {
+      mtx_unlock(&kq->kq_lock);
+      return EINTR;
+    }
+
+    if (tsp)
+      timeout = sleepts - getsystime();
+  }
+
+  /* To ensure the correctness of the iteration over pending events,
+   * we need to move already processed knotes to another list. `kq_head`
+   * can change in between iterations, because we are releasing the lock
+   * for a moment.
+   */
+
+  while (count < nevents) {
+    assert(mtx_owned(&kq->kq_lock));
+
+    if (TAILQ_EMPTY(&kq->kq_head))
+      break;
+
+    kn = TAILQ_FIRST(&kq->kq_head);
+    TAILQ_REMOVE(&kq->kq_head, kn, kn_penlink);
+
+    /* To call `filt_event` we need to lock `kn_objlock`, but firstly
+     * we need to release `kq_lock` since holding it would violate the locking
+     * order. `kn_objlock` must be taken before `kq_lock`.
+     */
+    mtx_unlock(&kq->kq_lock);
+    mtx_lock(kn->kn_objlock);
+    event = kn->kn_filtops->filt_event(kn, 0);
+    mtx_lock(&kq->kq_lock);
+    mtx_unlock(kn->kn_objlock);
+
+    /* Make sure that the event is still active. */
+    if (event == 0) {
+      kn->kn_status &= ~KN_QUEUED;
+      kq->kq_count--;
+      continue;
+    }
+
+    TAILQ_INSERT_HEAD(&knqueue, kn, kn_penlink);
+
+    eventlist[count++] = kn->kn_kevent;
+  }
+
+  TAILQ_CONCAT(&kq->kq_head, &knqueue, kn_penlink);
+
+done:
+  mtx_unlock(&kq->kq_lock);
+  *retval = count;
+
+  return 0;
+}
+
+static int kevent(proc_t *p, kqueue_t *kq, kevent_t *changelist,
+                  size_t nchanges, kevent_t *eventlist, size_t nevents,
+                  timespec_t *timeout, int *retval) {
+  int error;
+  size_t nerrors = 0;
+  kevent_t *kev;
+  void *obj;
+
+  /* First, make changes to the kqueue as requested. */
+  for (size_t i = 0; i < nchanges; i++) {
+    kev = &changelist[i];
+
+    error = kqueue_get_obj(p, kev, &obj);
+
+    if (error == 0 && (error = kqueue_register(kq, kev, obj)) == 0)
+      continue;
+
+    /* No more space left in the the eventlist, so simply return the error. */
+    if (nevents == nerrors)
+      return error;
```

Comment:
Shouldn't `retval` be set here, just like it is in line 356?

Reply:
No, this was a deliberate decision. 

From the kevent(2) manual:
```
If an error occurs while
     processing	an element of the changelist and there is enough room in the
     eventlist,	then the event will be placed in the eventlist with EV_ERROR
     set in flags and the system error in data.	 Otherwise, -1 will be re-
     turned, and errno will be set to indicate the error condition.
```

Reply:
Ok, thanks for clarifying that.